clientcontroller/opstackl2/consumer.go

```diff
@@ -264,13 +271,31 @@ func (cc *OPStackL2ConsumerController) SubmitBatchFinalitySigs(
 }
 
 // QueryFinalityProviderHasPower queries whether the finality provider has voting power at a given height
-// This interface function only used for checking if the FP is eligible for submitting sigs.
-// Now we can simply hardcode the voting power to true
-// TODO: see this issue https://github.com/babylonlabs-io/finality-provider/issues/390 for more details
 func (cc *OPStackL2ConsumerController) QueryFinalityProviderHasPower(fpPk *btcec.PublicKey, blockHeight uint64) (bool, error) {
+	pubRand, err := cc.QueryLastPublicRandCommit(fpPk)
+	if err != nil {
+		return false, err
+	}
+	if pubRand == nil {
+		return false, nil
+	}
+
+	// fp has 0 voting power if there is no public randomness at this height
+	lastCommittedPubRandHeight := pubRand.EndHeight()
+	cc.logger.Debug(
+		"FP last committed public randomness",
+		zap.Uint64("height", lastCommittedPubRandHeight),
+	)
+	if blockHeight > lastCommittedPubRandHeight {
```

Comment:
this is good but there is another case that this check won't cover

what if pubrand is not consecutive? so block 1-100 and 300-400 has PRs. if `blockHeight` is 200, `lastCommittedPubRandHeight` will be 400 and it should be considered not having VP

but this check doesn't cover it

Reply:
Good point as Babylon allows gaps. Better to add a TODO here.

Also, currently BTC-timestamping is not involved, right?

Reply:
yes it doesn't need to be in this PR. we can add a TODO

no currently BTC ts is not involved